server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java

```diff
@@ -128,7 +129,9 @@ public ThreadContext(Settings settings) {
         this.threadLocal = ThreadLocal.withInitial(() -> DEFAULT_CONTEXT);
         this.maxWarningHeaderCount = SETTING_HTTP_MAX_WARNING_HEADER_COUNT.get(settings);
         this.maxWarningHeaderSize = SETTING_HTTP_MAX_WARNING_HEADER_SIZE.get(settings).getBytes();
-        this.propagators = new CopyOnWriteArrayList<>(List.of(new TaskThreadContextStatePropagator()));
+        this.propagators = new CopyOnWriteArrayList<>(
+            List.of(new TaskThreadContextStatePropagator(), new QueryGroupThreadContextStatePropagator())
```

Comment:
Can we invoke `registerThreadContextStatePropagator` and register `QueryGroupThreadContextStatePropagator` instead of creating this dependency?  Similar to `ThreadContextBasedTracerContextStorage`

Reply:
We will have two choices to do that.
1. Move the registration to `Node` class.
2. Move the registration to `QueryGroupService` class which is going to be different PR.

Given the `Node` class is already pretty polluted I would rather go with the second approach,
What do you think @jainankitk @mch2 ?

Reply:
Agree I think putting this in the feature's own service is more appropriate.

Reply:
+1 to second.